io.openems.edge.timeofusetariff.entsoe/src/io/openems/edge/timeofusetariff/entsoe/TouEntsoeImpl.java

```diff
@@ -0,0 +1,182 @@
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
+
+import com.google.common.collect.ImmutableSortedMap;
+
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.utils.ThreadPoolUtils;
+import io.openems.edge.common.channel.value.Value;
+import io.openems.edge.common.component.AbstractOpenemsComponent;
+import io.openems.edge.common.component.OpenemsComponent;
+import io.openems.edge.common.currency.Currency;
+import io.openems.edge.common.meta.Meta;
+import io.openems.edge.timeofusetariff.api.TimeOfUsePrices;
+import io.openems.edge.timeofusetariff.api.TimeOfUseTariff;
+import io.openems.edge.timeofusetariff.api.utils.TimeOfUseTariffUtils;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "TimeOfUseTariff.ENTSO-E", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE //
+)
+public class TouEntsoeImpl extends AbstractOpenemsComponent implements TouEntsoe, OpenemsComponent, TimeOfUseTariff {
+
+	private static final int API_EXECUTE_HOUR = 14;
+
+	private final Logger log = LoggerFactory.getLogger(TouEntsoeImpl.class);
+	private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+	private final AtomicReference<ImmutableSortedMap<ZonedDateTime, Float>> prices = new AtomicReference<>(
+			ImmutableSortedMap.of());
+	private final List<Consumer<Value<Integer>>> subscriptions = new ArrayList<>();
+
+	@Reference
+	private Meta meta;
+
+	private Config config = null;
+	private Currency currency;
+	private ZonedDateTime updateTimeStamp = null;
+
+	public TouEntsoeImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				TouEntsoe.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+
+		if (!config.enabled()) {
+			return;
+		}
+		this.config = config;
+
+		Consumer<Currency> updateCurrency = currency -> {
+			this.currency = currency;
+			this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
+		};
+
+		// Trigger to activate when the currency from Meta is updated.
+		this.subscribe(updateCurrency);
+
+		updateCurrency.accept(this.meta.getCurrencyChannel().getNextValue().asEnum());
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+		this.unsubscribeAll();
+		ThreadPoolUtils.shutdownAndAwaitTermination(this.executor, 0);
+	}
+
+	/**
+	 * Subscribes to the Currency channel to trigger on update.
+	 * 
+	 * @param updateCurrency The callback {@link Consumer}.
+	 */
+	private void subscribe(Consumer<Currency> updateCurrency) {
```

Comment:
All this subscribe and unsubscribe is ok now, but it's still quite over engineered.

It would be sufficient to keep a private final variable:
```
private final Consumer<Currency> updateCurrency = currency -> {
			this.currency = currency;
			this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
		};
```

then subscribe during `activate()` (or `setMeta()`) and unsubscribe during `deactivate()` (or `unsetMeta()`). No need to keep a List of subscribes when you know you have only one.